Extend scope of "python.tensorBoard.logDirectory" and support variable substitution

evanphilip opened the pull request:
Fixes https://github.com/microsoft/vscode-jupyter/issues/8802

Credit: @PillarsZhang [patch-by-zhang](https://github.com/PillarsZhang/vscode-python/tree/patch-by-zhang)

evanphilip commented:
Bringing [this](https://github.com/microsoft/vscode-python/pull/20103) pull request to the attention of @DonJayamanne since issue [8802](https://github.com/microsoft/vscode-jupyter/issues/8802) is currently assigned to them. 

PS: I apologize for the spam if GitHub already handles the relevant notifications.

evanphilip commented:
- I’ve updated the pull request to pass the prettier linting check.

- I believe this scenario is already covered, at least partially, by the existing test in ‘ tensorBoardSession.test.ts’

karthiknadig commented:
@DonJayamanne Can you review this?

DonJayamanne commented on src/client/common/configSettings.ts:
```diff
@@ -504,7 +504,22 @@ export class PythonSettings implements IPythonSettings {
+            pythonSettings.get<ITensorBoardSettings>('tensorBoard'),
+        )!;
+        if (this.tensorBoard) {
+            Object.assign<ITensorBoardSettings, ITensorBoardSettings>(this.tensorBoard, tensorBoardSettings);
```
I'm not sure I understand the purpose of this?
Can we write this as follows?
```typescript
const tensorBoardSettings = systemVariables.resolveAny(
            pythonSettings.get<ITensorBoardSettings>('tensorBoard'),
        )!;
this.tensorBoard = tensorBoardSettings || { logDirectory: '' };
```

DonJayamanne commented on src/client/common/configSettings.ts:
```diff
@@ -504,7 +504,22 @@ export class PythonSettings implements IPythonSettings {
+            : {
+                  logDirectory: '',
+              };
+        const logDirectory = systemVariables.resolveAny(this.tensorBoard.logDirectory);
```
Isn't the previous `resolveAny` sufficient?

DonJayamanne commented on src/client/common/types.ts:
```diff
@@ -205,7 +205,7 @@ export interface IPythonSettings {
 
 export interface ITensorBoardSettings {
-    readonly logDirectory: string | undefined;
+    logDirectory: string;
```
I think we should leave the signature `string | undefined`, rather than initializing with an empty value

DonJayamanne reviewed (changes requested):
Thanks for the PR, just require some minor tweeks.

evanphilip commented:
@DonJayamanne I have incorporated your suggestions and simplified `configSettings.ts` significantly. I also stopped initializing `logDirectory` with an empty value. 

Thanks for your suggestions. I have never used TS before, so when in doubt, I copied the style from other parts of the script and made it unnecessarily complicated.

evanphilip commented:
@karthiknadig I have addressed the review comments by @DonJayamanne . Just wanted to make sure there isn’t anything else I need to do.

karthiknadig commented:
@DonJayamanne This one will need approval from you as well.
